src/drivers/magnetometer/lis2mdl/lis2mdl_i2c.cpp

```diff
@@ -0,0 +1,148 @@
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file lis2mdl_i2c.cpp
+ *
+ * I2C interface for LIS2MDL
+ */
+
+#include <px4_platform_common/px4_config.h>
+
+#include <assert.h>
+#include <debug.h>
+#include <errno.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include <string.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include <drivers/device/i2c.h>
+#include <drivers/drv_mag.h>
+#include <drivers/drv_device.h>
+
+#include "board_config.h"
+#include "lis2mdl.h"
+
+#define LIS2MDLL_ADDRESS        0x1e
+
+class LIS2MDL_I2C : public device::I2C
+{
+public:
+	LIS2MDL_I2C(int bus, int bus_frequency);
+	virtual ~LIS2MDL_I2C() = default;
+
+	virtual int     ioctl(unsigned operation, unsigned &arg);
+	virtual int     read(unsigned address, void *data, unsigned count);
+	virtual int     write(unsigned address, void *data, unsigned count);
+
+protected:
+	virtual int     probe();
+
+};
+
+device::Device *
+LIS2MDL_I2C_interface(int bus, int bus_frequency);
+
+device::Device *
+LIS2MDL_I2C_interface(int bus, int bus_frequency)
+{
+	return new LIS2MDL_I2C(bus, bus_frequency);
+}
+
+LIS2MDL_I2C::LIS2MDL_I2C(int bus, int bus_frequency) :
+	I2C("LIS2MDL_I2C", nullptr, bus, LIS2MDLL_ADDRESS, bus_frequency)
+{
+	_device_id.devid_s.devtype = DRV_MAG_DEVTYPE_LIS2MDL;
+}
+
+int
+LIS2MDL_I2C::ioctl(unsigned operation, unsigned &arg)
+{
+	switch (operation) {
+
+	case MAGIOCGEXTERNAL:
+		return external();
+
+	case DEVIOCGDEVICEID:
+		return CDev::ioctl(nullptr, operation, arg);
+
+	default:
+		return  -EINVAL;
+	}
+}
+
+int
+LIS2MDL_I2C::probe()
```

Comment:
Note for later, this is an example of the pattern we need to get right to cleanup the multi-interface driver mess. The probe shouldn't be specific to the interface.